internal/storage/sql/common/rule.go

```diff
@@ -430,8 +430,55 @@ func (s *Store) orderRules(ctx context.Context, runner sq.BaseRunner, namespaceK
 	return nil
 }
 
+func (s *Store) distributionValidationHelper(ctx context.Context, distributionRequest interface {
+	GetFlagKey() string
+	GetNamespaceKey() string
+	GetVariantId() string
+	GetRuleId() string
+}) error {
+	var ruleNamespace, variantNamespace, flagNamespace string
+
+	if err := s.builder.Select("namespace_key").
+		From("rules").
+		Where(sq.Eq{"id": distributionRequest.GetRuleId()}).
+		QueryRowContext(ctx).
+		Scan(&ruleNamespace); err != nil {
+		return err
+	}
+
+	if err := s.builder.Select("namespace_key").
+		From("variants").
+		Where(sq.Eq{"id": distributionRequest.GetVariantId()}).
+		QueryRowContext(ctx).
+		Scan(&variantNamespace); err != nil {
+		return err
+	}
+
+	if err := s.builder.Select("namespace_key").
+		From("flags").
+		Where(sq.And{sq.Eq{"\"key\"": distributionRequest.GetFlagKey()}, sq.Eq{"namespace_key": distributionRequest.GetNamespaceKey()}}).
+		QueryRowContext(ctx).
+		Scan(&flagNamespace); err != nil {
+		return err
+	}
+
+	// We return the sql.ErrNoRows here so that upstream we can return a more appropriate
+	// error to the client.
+	//nolint:gocritic
+	valid := (ruleNamespace == variantNamespace) && (ruleNamespace == flagNamespace)
+	if !valid {
+		return sql.ErrNoRows
+	}
+
+	return nil
```

Comment:
Open question for discussion: We could do this with a single query and a couple joins.

I set out this morning to try put one together and that is why I found the migration bug in sqlite 😂 .

```suggestion
	var count int

	if err := s.builder.Select("COUNT(*)").
		From("rules").
		Join("variants USING (namespace_key)").
		Join("flags USING (namespace_key)").
		Where(sq.Eq{
			"namespace_key": distributionRequest.GetNamespaceKey(),
			"rules.id": distributionRequest.GetRuleId(),
			"variants.id": distributionRequest.GetVariantId(),
			"flags.\"key\"": distributionRequest.GetFlagKey(),
		}).
		QueryRowContext(ctx).
		Scan(&count); err != nil {
		return err
	}
	
	if count < 1 {
		return sql.ErrNoRows
	}

	return nil
```

I did a similar query locally against `sqlite` and I generated a plan to check it wasn't doing anything wild. It appears to (in `sqlite`) appropriately use index scans:
```
sqlite> explain query plan select count(*) from rules join variants using(namespace_key) join flags using(namespace_key) where namespace_key = 'default' and rules.id='a13e9054-aef9-4fd5-83f8-38c57436cdab' and variants.id='749a21e7-c57d-418d-af91-be2b79a617d9' and flags.key='test';
QUERY PLAN
|--SEARCH rules USING INDEX sqlite_autoindex_rules_1 (id=?)
|--SEARCH variants USING INDEX sqlite_autoindex_variants_1 (id=?)
`--SEARCH flags USING COVERING INDEX sqlite_autoindex_flags_1 (namespace_key=? AND key=?)
```